File: LeetCode/Medium/0046. Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        int curr = nums[0];
        if(nums.size() == 1) {
            vector<int> one_perm = {curr};
            ans.push_back(one_perm);
            return ans;
        }

        vector<int> new_nums;
        for(int i = 1; i < nums.size(); i++) new_nums.push_back(nums[i]);
        vector<vector<int>> old_perm = permute(new_nums);

        for(int i = 0; i < old_perm.size(); i++) {
            for(int j = 0; j <= old_perm[i].size(); j++) {
                vector<int> one_perm = old_perm[i];
                one_perm.insert(one_perm.begin() + j, curr);
                ans.push_back(one_perm);
            }
        }
        return ans;
    }
};

File: LeetCode/Medium/0046. Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]

        idea:
         - for each iter i, use nums[i] as 1st elem
         - take the remaining elems to get the remaining perms
        """

        if len(nums) == 1: return [nums]

        ans = []
        for i in range(len(nums)):
            new_nums = nums[:i] + nums[i+1:]
            remaining_perms = self.permute(new_nums)

            for perm in remaining_perms:
                ans.append([nums[i]] + perm)
        
        return ans



        
